### ISSUE TITLE
Segmentation fault when cat list with all-empty cuda tensors

### ISSUE THREAD

**Issue by fmassa** [pytorch/pytorch#5739]
```python
import torch

a = torch.Tensor([]).cuda()
print(torch.cat([a]))  # segmentation fault
```

gdb points that the error might be in trying to get a size from an empty tensor:
```
(gdb) bt
#0  0x00007fffdf3b5364 in THCudaTensor_size () from /private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/lib/libATen.so.1
#1  0x00007fffdd16b4b5 in THCudaTensor_catArray () from /private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/lib/libATen.so.1
#2  0x00007fffdec8271f in at::CUDAFloatType::_cat(at::ArrayRef<at::Tensor>, long) const () from /private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/lib/libATen.so.1
#3  0x00007fffdeb11664 in at::native::cat(at::ArrayRef<at::Tensor>, long) () from /private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/lib/libATen.so.1
#4  0x00007fffded7a2fc in at::Type::cat(at::ArrayRef<at::Tensor>, long) const () from /private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/lib/libATen.so.1
#5  0x00007fffee4c3cbc in torch::autograd::VariableType::cat (this=0x55555657a858, tensors=..., dim=0) at torch/csrc/autograd/generated/VariableType.cpp:18026
#6  0x00007fffee6068c1 in at::cat (dim=0, tensors=...) at /private/home/fmassa/github/pytorch/torch/lib/tmp_install/include/ATen/Functions.h:2795
#7  torch::autograd::dispatch_cat (dim=0, tensors=...) at torch/csrc/autograd/generated/python_torch_functions_dispatch.h:717
#8  torch::autograd::THPVariable_cat (self=<optimized out>, args=<optimized out>, kwargs=<optimized out>) at torch/csrc/autograd/generated/python_torch_functions.cpp:1066
#9  0x0000555555662fd4 in _PyCFunction_FastCallDict ()
#10 0x00005555556f0bec in call_function ()
#11 0x000055555571519a in _PyEval_EvalFrameDefault ()
#12 0x00005555556eb529 in PyEval_EvalCodeEx ()
#13 0x00005555556ec2cc in PyEval_EvalCode ()
#14 0x0000555555768af4 in run_mod ()
#15 0x0000555555768ef1 in PyRun_FileExFlags ()
#16 0x00005555557690f4 in PyRun_SimpleFileExFlags ()
#17 0x000055555576cc28 in Py_Main ()
#18 0x000055555563471e in main ()
```

I'm using PyTorch version 0.4.0a0+66547ca

**Comment by zou3519** [comment]
There's a check that should exclude zero-dim tensors from `cat` (`a.dim()` should be 0 in this case), so I'm wondering why that's not happening right now...

**Comment by zou3519** [comment]
Related: https://github.com/pytorch/pytorch/issues/5332
We should probably rewrite torch.cat to better handle these cases

**Comment by zou3519** [comment]
Okay, I found the bug. The CUDA version of check doesn't check the case where the input contains all empty tensors, while the CPU version does. I'll put up a fix soon.
